pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -226,3 +230,103 @@ func buildGatewayInboundHTTPRouteConfig(env model.Environment, gatewayName strin
 		VirtualHosts: virtualHosts,
 	}
 }
+
+// buildGatewayInboundNetworkFilters retrieves all VirtualServices bound to the set of Gateways for this workload, filters
+// them by this server's port and hostnames, and produces network filters for each destination from the filtered services
+func buildGatewayInboundNetworkFilters(env model.Environment, server *networking.Server, gatewayNames []string) []listener.Filter {
+	dests := filterTCPDownstreams(env, server, gatewayNames)
+	filters := make([]listener.Filter, 0, len(dests))
+	for _, dest := range dests {
+		svc, err := env.GetService(dest.Host)
+		if err != nil {
+			log.Debugf("no service for host %q, skipping at ingress: %v", err)
+			continue
+		}
+		portNames := svc.Ports.GetNames()
+		instances, err := env.Instances(svc.Hostname, portNames, model.LabelsCollection{})
+		if err != nil {
+			log.Debugf("failed to retrieve instances of %s:%v with err: %v", svc.Hostname, portNames, err)
+			continue
+		}
+		for _, instance := range instances {
+			filters = append(filters, buildInboundNetworkFilter(instance))
```

Comment:
I apologize for the miscommunication earlier. I realized that the gateway.server has a full port spec with the protocol declared. We should only build the filter stack for the protocol declared at the server and not bother with the protocols associated with backend instances. In other words, this entire forloop is unnecessary.

Reply:
second bug: you need to buildoutboundNetworkFilters.. gateway is just a single proxy, with no backing app in the pod.. So, essentially, what hits gateway goes out immediately, i.e. its a outbound sidecar thats standalone. 

You need outboundNetworkFilters because you need to use the subsetnames when building the cluster from the gateway.

You only use inbound for service instances. Everything else essentially uses outbound (gateay/sidecar)